server/game/gameSystems/DelayedEffectSystem.ts

```diff
@@ -0,0 +1,93 @@
+import AbilityHelper from '../AbilityHelper';
+import { AbilityContext } from '../core/ability/AbilityContext';
+import { IAbilityLimit } from '../core/ability/AbilityLimit';
+import { TriggeredAbilityContext } from '../core/ability/TriggeredAbilityContext';
+import { Duration, EventName } from '../core/Constants';
+import { GameEvent } from '../core/event/GameEvent';
+import { GameSystem, IGameSystemProperties } from '../core/gameSystem/GameSystem';
+import { WhenType } from '../Interfaces';
+import * as Contract from '../core/utils/Contract';
+
+export enum DelayedEffectType {
+    Card,
+    Player
+}
+
+export interface IDelayedEffectSystemProperties extends IGameSystemProperties {
+    title: string;
+    when: WhenType;
+    duration?: Duration;
+    limit?: IAbilityLimit;
+    immediateEffect: GameSystem<TriggeredAbilityContext>;
+    effectType: DelayedEffectType;
+}
+
+export class DelayedEffectSystem<TContext extends AbilityContext = AbilityContext> extends GameSystem<TContext, IDelayedEffectSystemProperties> {
+    public override readonly name: string = 'applyDelayedEffect';
+    public override readonly eventName: EventName = EventName.OnEffectApplied;
+    public override readonly effectDescription: string = 'apply a delayed effect';
+
+    protected override defaultProperties: IDelayedEffectSystemProperties = {
+        title: null,
+        when: null,
+        duration: Duration.Persistent,
+        limit: AbilityHelper.limit.perGame(1),
+        immediateEffect: null,
+        effectType: null
+    };
+
+    public eventHandler(event: any, additionalProperties: any): void {
+        const properties = this.generatePropertiesFromContext(event.context, additionalProperties);
```

Comment:
I missed this earlier - I realized that the design intent for these systems is that we shouldn't access properties directly here but instead put everything relevant on the event and then read from there when we get to eventHandler. I'm planning on slowly working my way through and fixing that in older systems.

Below comments suggest overriding addPropertiesToEvent anyway so we could just add everything needed there

Reply:
So we would want to move the switch logic here to that method? Would I even need the eventHandler method anymore?

Reply:
Most of the logic can stay here, but anything you would read off of properties should instead be added to the event via `addPropertiesToEvent`